### ISSUE TITLE
Question: is a thread per node really necessary?

### ISSUE THREAD

**Issue by facontidavide** [miccol/Behavior-Tree#9]
Hi,

I noticed that a single thread is created for each node. This seems like a waste of resources in large systems with potentially dozens of nodes.

I guess that either a custom (simple) scheduler or a coroutine library might be able to do the job.
Considering the overhead of the context switching, I have the feeling that even traversing the entire tree might be faster.

Is there any reason I am not aware of to use a separate thread per node?

Cheers

Davide

**Comment by miccol** [comment]
Hi Davide,

The reason behind the thread per node is that a running node should run in a separate thread from its parent. This is because the HALT of a node may be requested any time, asynchronously from the running node. For instance, imagine the tree: Sequence(Condition, Action). We need a thread that executes the action (Action Thread) and a thread that verify if the action must be halted (Sequence Thread).

Probably there is a better approach to this, but for my research I "just" make proof-of-concept BTs.

Do you have any suggestion?

Thank you

Michele

**Comment by facontidavide** [comment]
Hi Michele,

I am not sure I see the problem the way you do.

1) At the end of the day, your threads need to explicitly check the condition variable and interrupt their while() loop. From the point of view of the signal they work as "cooperative" multitsking.

2) The only node where there MIGHT be a need for threads are ActionNodes. It is very hard for me to see how a ControlNode might nee a separate thread. The fact that they shall be able to propagate the signal has nothing to do with the fact that they own their own thread.

3) You can serialize the Exec() execution using boost::asio, but this require some thinking, I guess.

If you don't mind, I will play with it on my own fork and ply with different ideas I have.

Davide

**Comment by facontidavide** [comment]
Let's take a look to [ActionNodeExample](https://github.com/miccol/Behavior-Tree/blob/master/src/ActionNodeExample.cpp)

A) We are waiting for the tick here.  https://github.com/miccol/Behavior-Tree/blob/master/src/ActionNodeExample.cpp#L23
This means that the following code is execvuted, oif course, once per tick. Ideally, we could just remove the thread and the while(true) and delegate the execution of the code inside Exec() to a centralized scheduler.

B) Despite the complex infrastructure of signals and waits, the user MUST actively check for interruption here:
https://github.com/miccol/Behavior-Tree/blob/master/src/ActionNodeExample.cpp#L41
Otherwise halt() has no effect.
I am thinking: if we remove the assumption that multiple threads are necessary, this piece of code is still the same.

C) I must admit that from the user point of view, the logic of halting and state trnasitions is pretty complicated. For example I don't understand the need for this wait
https://github.com/miccol/Behavior-Tree/blob/master/src/ActionNodeExample.cpp#L78

**Comment by miccol** [comment]
Sure, please feel free to improve it.

Just make sure that also the control flow nodes are preemptable (or find a smarter way to traverse the tree), probably the solution of one thread per node is a bit overkill. If you find a better one I will adopt it.


Thank you.

Michele

P.S. the reason why I used boost is because back then ROS could not support C++11.

**Comment by facontidavide** [comment]
I think I have something you can look at here:

https://github.com/facontidavide/BehaviorTree/tree/refactoring

__WARNING:__ I need to do unit testing, so the code might still contain severe bugs. Just focus on the concept and the API.

The main idea is that none of the nodes has by default its own thread.
The only one that has a separate thread is potentially the ActionNode. I created a thin wrapper that makes this case easier to manage, called __ActionNodeAsync__

https://github.com/facontidavide/BehaviorTree/blob/refactoring/lib/include/ActionNode.h#L27

To get an idea of how the user code shall be written, take a look here 

https://github.com/facontidavide/BehaviorTree/blob/refactoring/sample/src/ActionNodeExample.cpp

As you can see it is dramatically simpler, since exec() implementation shall not focus on the different BT states and the logic of the signal, that was quite hard to follow IMHO.

What do you think? Does it make sense or am I missing something?

__PS__: The halt "signal" just propagates through the tree using plain old recursion.
__PPS__: C++11 is not related to the ROS version but rather to the compiler used by the ROS version installed on your PC. Therefore as long as we use functionalities supported by gcc 4.8, we should be fine.

**Comment by miccol** [comment]
Thank you. I will check ASAP (bit busy in this period)

(ROS hydro did not support C++11, to compile in ROS you have to use catkin http://answers.ros.org/question/12227/ros-and-c11/)

Michele

**Comment by facontidavide** [comment]
ok, I was thinking about Indigo or later.

I am still doing many changes to my proposal, but I am quite confident that this approach would simplify considerably the code, from both the point of view of the library or the user code (and reduce the use of resources and latency).

this is a sample of the users' code 100% equivalent to yours in terms of functionalities

```c++
ActionNodeExample::ActionNodeExample(const std::string& name) :
    ActionNodeAsync(name),
    _max_cycles(1),
    _halt_requested(false)
{ }

BT::State ActionNodeExample::exec()
{
    // user must proactively split long computation and check
    // _halt_requested to see if the onHalt callback was executed

    for (int cycle_count=0; cycle_count < _max_cycles; cycle_count++)
    {
        // if _halt_requested is true, it is reset atomically to false.
        bool pending_halt = true;
        if( _halt_requested.compare_exchange_strong( pending_halt, false )) {
            // _halt_requested was true
            return ( BT::State::IDLE);
        }
        std::cout << name() << " working " << cycle_count+1
                       << "/" << _max_cycles << std::endl;
        std::this_thread::sleep_for( std::chrono::milliseconds(350));
    }
    // if you got here, it means that (_cycle_count == _time)
    // and no halt was received.
    return( BT::State::SUCCESS );
}

void ActionNodeExample::setCyclesCount(int cycles) {
    this->_max_cycles = cycles;
}

void ActionNodeExample::onHalt() {
    // the thread calling this callback almost surely is NOT the same
    // that runs exec().
    // For this reason you must change _halt_requested in a thread safe way,
    // using either a mutex or an atomic.
    _halt_requested.store(true);
}
```

**Comment by miccol** [comment]
I had a quick look (I have a strict deadline soon),

looks cool but if you create a SequenceNode composition (not SequenceStarNode) of a ConditionNodeExample with a ActionNodeExample setting the max cycles to say 1000 when the condition node returns failure the ActionNodeExample keeps running.

Can you try that?

**Comment by facontidavide** [comment]
As I said, the implementation is still being tested and it is not ready for prime time.

Nevertheless you were right, it was not working. I fixed it and added a unit test:

https://github.com/facontidavide/BehaviorTree/blob/refactoring/tests/bt_async_sample.cpp

If you are busy, you probably want to postpone a detailed review to the next week, since I am still testing the corner cases.

**Comment by miccol** [comment]
Unfortunately I will be rather busy until March (I am writing my PhD thesis.) We can continue the discussion over email if you prefer (miccol@kth.se).

Anyway your way of programming looks much better I have to say. I did not have much experience in multithreading programming when I started developing the library but for my research worked OK (I and do not have time to code as I wish), in the ROS implementation I have released the  version 1.2 where I have applied the Google Code style and improved the tree visualization.  I am planning to release the version 2.0 after my defence. I will probably adopt you tick traversal method, giving you the credit of course.

**Comment by facontidavide** [comment]
Good luck with your thesis :)

Once you decide to work on 2.0. drop me a line and we can merge our two libraries and collaborate on a single one.

In the following weeks I will start using my own version in real-world projects and I will validate the  model.

Cheers
